src/compiler/checker.ts

```diff
@@ -44,6 +44,7 @@ namespace ts {
 
         let compilerOptions = host.getCompilerOptions();
         let languageVersion = compilerOptions.target || ScriptTarget.ES3;
+        var modulekind = compilerOptions.module ? compilerOptions.module : languageVersion === ScriptTarget.ES6 ? ModuleKind.ES6 : ModuleKind.None;

```

Comment:
> Perhaps it would be better to remove ModuleKind.ES6, as it could be confusing to API consumers, and instead use this here

_Please_ no. Not having ES6 as a module kind is all kinds of not forward compatible and causes odd checks which depend on there not being module kinds alongside specific language versions. Going forward, when we implement #4813 those forms of exports will _not_ be ES6 compatible module emit, but _can_ be compiled _to_ ES6 compatible module emit. In effect, when we add an ES7 target and ES7 compatible module emit feature, we're going to _need_ an ES6 module kind (and, to be consistent, hopefully an ES7 module kind!). Additionally, not having an ES6 module kind makes specifying that you want to emit an ES6 module style bundle for #4754  awkward. (Do I target ES6 but not specify module output type at all? That's oddly inconsistent.)

`--module` is one of our few variable feature flags at present, it's time we start treating it as such. ES6 is a kind of module emit. We should treat it as such. The module format is part of the ES6 spec, true, but nothing inherently ties the feature to the ES6 target only. Right now it seems nonsensical to ask for ES6 module syntax with a target below ES6 (since we know of 0 runtimes which support ES6 modules and yet support no other ES6 features), but we certainly _could_ emit it.

IMO, having an ES6 module kind _simplifies_ our API. When a module kind is set, it now controls the emit for the module format. _Full stop._